Add order api

mhdtahawi opened the pull request.

mhdtahawi committed:
Update gradle version
Update .gitignore: add .idea directory
Remove .idea directory from the repository

mhdtahawi committed:
Add lombok plugin to build.gradle
Add InvalidStateException class for invalid input
Add ApiInput interface: to be extended by all input classes
Add AddStandardOrderInput and AddStandardOrderResult classes
Update KrakenApiClient: new method addStandardOrder

osenseijedi commented on build.gradle:
```diff
@@ -1,18 +1,22 @@
+version '1.0-SNAPSHOT'
+
+
+sourceCompatibility = 1.8
```
should be 1.11 from now on

mhdtahawi replied:
1.11 or just 11 ?

osenseijedi replied:
1.11

osenseijedi commented on build.gradle:
```diff
@@ -1,18 +1,22 @@
+}
+
+dependencies {
+    compile 'com.fasterxml.jackson.core:jackson-databind:2.8.9'
```
All deps should be "implementation".
Also should we upgrade all libs versions here too ?

osenseijedi commented on src/main/java/com/ptrader/connector/kraken/input/AddStandardOrderInput.java:
```diff
@@ -0,0 +1,289 @@
+        }
+    }
+
+    private AddStandardOrderInput(String pair, TransactionType type, double volume, boolean validate) {
```
double: we should use BigDecimal instead.

osenseijedi commented on src/main/java/com/ptrader/connector/kraken/input/AddStandardOrderInput.java:
```diff
@@ -0,0 +1,289 @@
+import java.util.stream.Collectors;
+
+@Getter
+public class AddStandardOrderInput implements ApiInput {
```
I would actually separate the object itself and its factory

mhdtahawi committed:
build.gradle: Set sourceCompatibility to 11
Update versions of libraries
Use testng instead of junit for testing
Introduce new factory class for AddStandardOrderInput
Use BigDecimal instead of Double in AddStandardOrderInput fields

mhdtahawi committed:
Rename the field validate to avoid confusion
Correctly write the value of validate (now validateInputOnly) to the result map in getInput method

mhdtahawi committed:
Factor out common parts of all AddStandardOrderInputFactory methods into one private method.

mhdtahawi committed:
Initialize all Optional<T> fields in AddStandardOrderInputFactory to empty.
